### PR TITLE
feat: deterministic builds

### PR DESCRIPTION
**Pull request by lrvick**
Use SOURCE_DATE_EPOCH, explicit timestamp setting, and the cpio
--reproducible flag in order to get hash-for-hash identical builds
across multiple systems.

Tested on Debian and Arch Linux systems

Signed-off-by: Lance R. Vick <lance@lrvick.net>

# Pull Request

## What? (description)

The generated build output of this repository is now deterministic across multiple systems.

## Why? (reasoning)

This resolves the only hard blocker to pursue the improved supply chain integrity recommendations proposed here: https://github.com/talos-systems/rfcs/pull/2

Notably multiple independently controlled systems should be able to reproduce published binaries to prove they have not been tampered with, provided the cached toolchain is not tampered with (future work).

## Acceptance

- [x] you linked an issue (if applicable)
- [x] you ran conformance (`make conformance`)
- [x] manually tested on multiple systems

### CODE REVIEW

**Review by andrewrynhard — COMMENTED** [review]
🎉

File: Dockerfile [hunk 1]
```diff
@@ -409,7 +412,14 @@ FROM build AS initramfs-archive-arm64
 WORKDIR /initramfs
 COPY --from=squashfs-arm64 /rootfs.sqsh .
 COPY --from=init-build-arm64 /init .
-RUN set -o pipefail && find . 2>/dev/null | cpio -H newc -o | xz -v -C crc32 -0 -e -T 0 -z >/initramfs.xz
+RUN find . -print0 \
```
Comment (smira) [inline Dockerfile:415]: Do we need apply same for amd64 variant?

Comment (lrvick) [reply Dockerfile:415]: Yeah this was just introduced with the rebase against master for me.

Also Found a one line diff regression as a result of that as well.

Grabbing both. Don't merge quite yet!

**Comment by lrvick** [comment]
Corrected to consider new split architecture blocks inherited from master rebase.

Re-testing.

**Comment by andrewrynhard** [comment]
@lrvick Is this ready to merge?

**Comment by lrvick** [comment]
Having a couple repro issues that are possibly cache related.

I will post my hashes shortly.

When someone other than me can reproduce, particularly someone on a
different OS than me, I say we call this good.

**Comment by andrewrynhard** [comment]
> Having a couple repro issues that are possibly cache related. I will post my hashes shortly. When someone other than me can reproduce, particularly someone on a different OS than me, I say we call this good.
> […](#)

Great, I can try to reproduce tonight.

**Comment by lrvick** [comment]
Okay, looks like my issue is due to tags that happened in the time between when I obtained my cached dependency builds on both systems.

```
 │ ├── strings --all --bytes=8 {}
│ │ │ │ │ @@ -1,8 +1,8 @@
│ │ │ │ │ -WyzmoaC2TfQ2rY3wM-uR/xxgAPRqhFtJnM_aTUJAt/No-cfyQsp1taCqBgNI89/NdK2NYZh7VmkJZOdWgQ-
│ │ │ │ │ +IcqjIwhBzDBH06ZjgFMe/xxgAPRqhFtJnM_aTUJAt/No-cfyQsp1taCqBgNI89/7J3oxvO9VS654ovOBt6r
│ │ │ │ │  UUUUUUUUH!
│ │ │ │ │  33333333H!
│ │ │ │ │  runtime L
│ │ │ │ │   error: L
│ │ │ │ │  =_B>fuFH
│ │ │ │ │  D$hH9B(t
│ │ │ │ │  UUUUUUUUH
│ │ │ │ │ @@ -65521,19 +65521,19 @@
│ │ │ │ │  UUUUUUUUUUUUUUUU
│ │ │ │ │  UUUUUUUUUUUUUUUU
│ │ │ │ │  UUUUUUUUUUUUUUUU
│ │ │ │ │  UUUUUUUUUUUUUUUU
│ │ │ │ │  UUUUUUUUUUUUUUUU
│ │ │ │ │  UUUUUUUUUUUUUUUU
│ │ │ │ │  UUUUUUUUUUUUUUUU
│ │ │ │ │ -c047fb89
│ │ │ │ │ +d5b88220
│ │ │ │ │  talos-systems
│ │ │ │ │  /toolchain/go
│ │ │ │ │  call frame too large
│ │ │ │ │ -v0.10.0-alpha.2-53-gc047fb89
│ │ │ │ │ +v0.10.0-alpha.2-51-gd5b88220
```

We will need some tooling that rebuilds all from 0 against fixed versions so you are always guaranteed to get the same result for a given commit. This will also solve the remaining trust problem of potential mutations to cached remote toolchains that we need to also be able to reproduce before we can make strong gurantees about tamper ressitance.

I will make a proposal to address both in one shot as a follow up, but I don't see it as a blocker for this PR.

Going to validate with two new builds from empty docker cache back to back now.

**Comment by lrvick** [comment]
Okay!

So once I am working from clean trees on two systems, one debian, one arch, I can run the following aggressive command to nuke all build cache and in fact get identical results.

```
$ docker system prune --volumes --all && docker buildx prune --all && make PLATFORM=linux/amd64
...
$ sha256sum _out/*

e16045ce9554cffabc66b4873f8d0a90b92a267a75b8e817eba6c8c7e2ddabe2  _out/initramfs-amd64.xz
e88444b55dfd9f76d2312ed6a9cea51922e2a1b507f9d536d4b9d4f3eb29bd2a  _out/talosctl-linux-amd64
9b5278d2cb6cc17c8d7354febeb561bb5ff424d51a61e56362b7ca43e211ca8e  _out/talosctl-linux-arm64
616edd2a6a3ed40ac78ccb0876ebfa78f5d2a8e32f0bd3ff7681a9d8d9aa5d9e  _out/talosctl-linux-armv7
05530658889ed82e2d0d4c1ef8e7c21b8827e0229634c3a6e2e5a06131b93d4a  _out/vmlinuz-amd64

```

Feel free to try and replicate these, otherwise I feel confident this works as expected on my side.

**Comment by andrewrynhard** [comment]
> Okay!
> 
> So once I am working from clean trees on two systems, one debian, one arch, I can run the following aggressive command to nuke all build cache and in fact get identical results.
> 
> ```
> $ docker system prune --volumes --all && docker buildx prune --all && make PLATFORM=linux/amd64
> ...
> $ sha256sum _out/*
> 
> e16045ce9554cffabc66b4873f8d0a90b92a267a75b8e817eba6c8c7e2ddabe2  _out/initramfs-amd64.xz
> e88444b55dfd9f76d2312ed6a9cea51922e2a1b507f9d536d4b9d4f3eb29bd2a  _out/talosctl-linux-amd64
> 9b5278d2cb6cc17c8d7354febeb561bb5ff424d51a61e56362b7ca43e211ca8e  _out/talosctl-linux-arm64
> 616edd2a6a3ed40ac78ccb0876ebfa78f5d2a8e32f0bd3ff7681a9d8d9aa5d9e  _out/talosctl-linux-armv7
> 05530658889ed82e2d0d4c1ef8e7c21b8827e0229634c3a6e2e5a06131b93d4a  _out/vmlinuz-amd64
> ```
> 
> Feel free to try and replicate these, otherwise I feel confident this works as expected on my side.

I can confirm that I get the same SHAs! Going to rebase before merging.

**Pull request merged**